### PR TITLE
feat: Qdrant as a supported knowledge base

### PR DESCRIPTION
**Pull request by Anush008**
## Description

This PR adds support for [Qdrant](https://qdrant.tech/) to be used as a knowledge base.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Test Plan

Tests have been added to `tests/system/qdrant` for the sync and async implementations and are configured to run using both, an in-memory Qdrant instance and Qdrant server(if running at port 6333, skipped otherwise).

## Note
The `qdrant_client` dependency can be gated behind an extra if needed and CLI support can be added subsequently.

### CODE REVIEW

**Review by ahmadmustafaanis — APPROVED** [review]

**Comment by igiloh-pinecone** [comment]
Thank you @Anush008 for the great contribution!!   
We are still in the process of reviewing this PR, will publish the review in the coming days.

**Comment by Anush008** [comment]
Thanks @igiloh-pinecone.

**Comment by ChristopherDosin** [comment]
> We are still in the process of reviewing this PR, will publish the review in the coming days.

Any news on that a month later? :)

**Comment by igiloh-pinecone** [comment]
> > We are still in the process of reviewing this PR, will publish the review in the coming days.
> 
> Any news on that a month later? :)

Hi @ChristopherDosin,  
As I mentioned before, we are considering an internal refactor of the `KnowledgeBase` class, that could  affect the approach taken in this PR. But since this is still in internal discussions and design process, we shouldn't delay merging this PR any longer.  
I'll publish a review of the current standing PR in the coming day or two.  
Thank you for you patience! 🙏

**Comment by ChristopherDosin** [comment]
> > > We are still in the process of reviewing this PR, will publish the review in the coming days.
> > 
> > 
> > Any news on that a month later? :)
> 
> Hi @ChristopherDosin, As I mentioned before, we are considering an internal refactor of the `KnowledgeBase` class, that could affect the approach taken in this PR. But since this is still in internal discussions and design process, we shouldn't delay merging this PR any longer. I'll publish a review of the current standing PR in the coming day or two. Thank you for you patience! 🙏

Awesome, thank you my friend - Just wondering which one performs better :)

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (igiloh-pinecone) [inline tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:35]: Not required. We have `pytest-dotenv` installed by default now

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (igiloh-pinecone) [inline tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:1]: @Anush008 There is a lot of code duplication between this file and `test_knowledge_base`, this is something we're trying very much to avoid.   
Please see if you can re-write this test suite to reuse the tests in `test_knowledge_base.py` as much as possible. When way to do this would be by parameterizing the `knowledge_base` fixture like we're doing for [LLM tests](https://github.com/pinecone-io/canopy/blob/50414f5e1a19d0ed86aa5edeece955ef7d5b44f7/tests/system/llm/test_openai.py#L63) or by creating inheritance relationship like we're doing for [Tokenizer tests](https://github.com/pinecone-io/canopy/blob/50414f5e1a19d0ed86aa5edeece955ef7d5b44f7/tests/system/tokenizer/test_cohere_api_tokenizer.py#L10).

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (igiloh-pinecone) [inline tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:7]: Since `qdrant_client` would become an extra dependency, this import will need to be changed to `pytest.importorskip` (or any alternative mechanism that will simply skip testing `QdrantKnowledgeBase` for users who haven't installed the `qdrant` extra).

**Review by igiloh-pinecone — CHANGES_REQUESTED** [review]
@Anush008 thank you very much for your contribution!!  

Apologies for the delayed review.  
Please see a few comments. Namely, please try to re-use as much `test_` code as possible instead of duplicating it. This will make future changes to `KnowledgeBase` much more maintainable.

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (Anush008) [reply tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:1]: Parameterizing using a fixture isn't straightforward because the Pinecone and Qdrant knowledge base classes have different constructor options.
I initially thought of inheriting the tests. But the tests in `test_knowledge_base.py` were functional and I didn't want to fiddle/edit with those.

File: src/canopy/knowledge_base/qdrant/qdrant_knowledge_base.py
Comment (Anush008) [reply src/canopy/knowledge_base/qdrant/qdrant_knowledge_base.py:664]: In Qdrant there's no built-in system for fusing the scores or a hybrid search.
But 2 separate searches with sparse and dense vector with a custom fusing strategy to be done by the caller.

File: src/canopy/knowledge_base/qdrant/constants.py
Comment (Anush008) [reply src/canopy/knowledge_base/qdrant/constants.py:1]: The constant is named `INDEX_NAME_PREFIX` in `knowledge_base.py` for Pinecone. Whereas it's a 'collection' for Qdrant. So I've declared them separately.
I can maybe do 
```python
from canopy.knowledge_base.knowledge_base import INDEX_NAME_PREFIX
COLLECTION_NAME_PREFIX = INDEX_NAME_PREFIX 
```
in this file to maintain the terminology for the Qdrant implementation.

File: src/canopy/knowledge_base/qdrant/converter.py
Comment (Anush008) [reply src/canopy/knowledge_base/qdrant/converter.py]: https://github.com/pinecone-io/canopy/pull/244/commits/e0275213f6aa38f085e2de1d3dd45945a75ece97

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (Anush008) [reply tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py]: https://github.com/pinecone-io/canopy/pull/244/commits/289577501bf0b1263172c3e22d8f3fb16ce2ee5f

File: src/canopy/knowledge_base/qdrant/converter.py
Comment (Anush008) [reply src/canopy/knowledge_base/qdrant/converter.py]: https://github.com/pinecone-io/canopy/pull/244/commits/289577501bf0b1263172c3e22d8f3fb16ce2ee5f

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (igiloh-pinecone) [reply tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:1]: @izellevy WDYT? Is there an easy way we can de-dupe this test?  
Should we merge as-is for now and deal with this later?

@miararoy your 2c would also be appreciated.

File: pyproject.toml
Comment (Anush008) [reply pyproject.toml]: https://github.com/pinecone-io/canopy/pull/244/commits/3f34b7a16a2e1729ce7769e5513f7463672662ce

File: src/canopy/knowledge_base/qdrant/qdrant_knowledge_base.py
Comment (Anush008) [reply src/canopy/knowledge_base/qdrant/qdrant_knowledge_base.py]: https://github.com/pinecone-io/canopy/pull/244/commits/3f34b7a16a2e1729ce7769e5513f7463672662ce

File: pyproject.toml
Comment (Anush008) [reply pyproject.toml]: https://github.com/pinecone-io/canopy/pull/244/commits/3f34b7a16a2e1729ce7769e5513f7463672662ce

File: src/canopy/knowledge_base/qdrant/constants.py
Comment (Anush008) [reply src/canopy/knowledge_base/qdrant/constants.py]: https://github.com/pinecone-io/canopy/pull/244/commits/3f34b7a16a2e1729ce7769e5513f7463672662ce

File: pyproject.toml
Comment (Anush008) [reply pyproject.toml]: https://github.com/pinecone-io/canopy/pull/244/commits/d9f80e7ad32ce987e8de3ef64cfb9bf374cd8325

File: tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py
Comment (igiloh-pinecone) [reply tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:1]: @Anush008 thanks for your patience!   
I'm approving. Let's merge!!

Comment (Anush008) [reply tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:1]: 🎊🎊

Comment (igiloh-pinecone) [reply tests/system/knowledge_base/qdrant/test_qdrant_knowledge_base.py:1]: @Anush008 the branch doesn't seem to be up to date with `main`. I confirm locally on my env that after merging `main` tests do pass on this branch.  
Please merge `main` so we can proceed with the merge

**Review by igiloh-pinecone — APPROVED** [review]
LGTM!

Thank you @Anush008 for this effort

**Comment by brightsparc** [comment]
Looking forward to having this land. Anything holding this up?

**Comment by igiloh-pinecone** [comment]
> Looking forward to having this land. Anything holding this up?

As far as I recall some system test failed in the merge queue 🤔 .  
Trying to re-merge this now, let's see.

**Comment by Anush008** [comment]
@igiloh-pinecone, I've updated the branch. Could you try now please?

**Comment by Anush008** [comment]
The flaky test on Python 3.9 seems to arise from an unrelated change.

**Comment by Anush008** [comment]
https://github.com/pinecone-io/canopy/pull/244#issuecomment-1997496521 should be resolved with #327.

**Comment by brightsparc** [comment]
Nice. When is the next release likely to get cut?

**Comment by izellevy** [comment]
We have a new release: https://github.com/pinecone-io/canopy/releases/tag/v0.9.0

**Pull request merged**
